src/crimson/os/seastore/nvmedevice/nvmedevice.cc

```diff
@@ -0,0 +1,227 @@
+            write_granularity = namespace_data.npwg + 1;
+            write_alignment = namespace_data.npwa + 1;
+            });
+      }
+      return seastar::now();
+      }).handle_error(
+        /* If device does not support ioctl, just open without stream */
+        crimson::ct_error::input_output_error::handle([this, in_path, mode](auto) {
+          open_for_io(in_path, mode);
+          return seastar::now();
+        }));
+}
+
+void
+NormalNBD::open_for_io(const std::string &in_path, seastar::open_flags mode)
+{
+  ctx.resize(write_life_max);
+  for (uint32_t i = 0; i < write_life_max; i++) {
+    stream_fd.push_back(seastar::file_desc::open(in_path, (int)mode));
+    if (i != write_life_not_set) {
+      int posix_fd = stream_fd[i].get();
+      fcntl(posix_fd, F_SET_FILE_RW_HINT, &i);
+    }
+
+    io_setup(MAX_EVENTS, &ctx[i]);
+  }
+  completion_poller = std::thread(poll_completion, &ctx, &exit);
+}
+
+write_ertr::future<>
+NormalNBD::write(
+  uint64_t offset,
+  bufferptr &bptr,
+  uint16_t stream) {
+  logger().debug(
+      "block: do_write offset {} len {}",
+      offset,
+      bptr.length());
+  io_context_t io_context = io_context_t();
+  io_prep_pwrite(
+      &io_context.cb,
+      stream_fd[stream].get(),
+      bptr.c_str(),
+      bptr.length(),
+      offset);
+  iocb* cb_ptr[1] = {&io_context.cb};
+  io_submit(ctx[stream], 1, cb_ptr);
+  return seastar::do_with(std::move(io_context), [] (auto& io_context) {
+    /*
+     * libaio needs additional poller thread (see poll_completion) to poll IO
+     * completion. When the poller catches a completion, it marks "done" field
+     * of corresponding io_context.
+     */
+    if (io_context.done) {
+      return seastar::now();
+    }
+      return seastar::later();
+    });
+}
+
+read_ertr::future<>
+NormalNBD::read(
+  uint64_t offset,
+  bufferptr &bptr) {
+  logger().debug(
+      "block: do_read offset {} len {}",
+      offset,
+      bptr.length());
+  io_context_t io_context = io_context_t();
+  io_prep_pread(
+      &io_context.cb,
+      stream_fd[0].get(),
+      bptr.c_str(),
+      bptr.length(),
+      offset);
+  iocb* cb_ptr[1] = {&io_context.cb};
+  io_submit(ctx[0], 1, cb_ptr);
+  return seastar::do_with(std::move(io_context), [] (auto& io_context) {
+      if (io_context.done) {
+      return seastar::now();
+      }
+      return seastar::later();
+      });
+}
+
+seastar::future<>
+NormalNBD::close() {
+  logger().debug(" close ");
+  exit = true;
+  completion_poller.join();
+  fd.close();
+  return seastar::now();
+}
+
+nvme_command_ertr::future<>
+NormalNBD::pass_through_io(NVMePassThroughCommand& command) {
+  logger().debug("block: pass through");
+  int ret = fd.ioctl(NVME_IOCTL_IO_CMD, command);
+  if (ret < 0) {
+    logger().debug("block: pass through failed");
+    return crimson::ct_error::input_output_error::make();
+  }
+  else {
+    return nvme_command_ertr::now();
+  }
+}
+
+nvme_command_ertr::future<>
```

Comment:
This formatting is inconsistent with the rest of seastore/.  I commented here, but I want you to go through and make everything match as much as possible.

Reply:
OK